http/engine.go

```diff
@@ -0,0 +1,313 @@
+}
+
+// Start starts the HTTP engine.
+func (h *Engine) Start() error {
+	go func(server *MultiEcho, cancel func()) {
+		if err := server.Start(); err != nil {
+			if !errors.Is(err, http.ErrServerClosed) {
+				log.Logger().
+					WithError(err).
+					Error("HTTP server stopped due to error")
+			}
+		}
+		cancel()
+	}(h.server, h.serverShutdownCb)
+	return nil
+}
+
+// Shutdown shuts down the HTTP engine.
+func (h *Engine) Shutdown() error {
+	return h.server.Shutdown(context.Background())
+}
+
+// decodeURIPath is echo middleware that decodes path parameters
+func decodeURIPath(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		// FIXME: This is a hack because of https://github.com/labstack/echo/issues/1258
+		newValues := make([]string, len(c.ParamValues()))
+		for i, value := range c.ParamValues() {
+			path, err := url.PathUnescape(value)
+			if err != nil {
+				path = value
+			}
+			newValues[i] = path
+		}
+		c.SetParamNames(c.ParamNames()...)
+		c.SetParamValues(newValues...)
+		return next(c)
+	}
+}
+
+// matchesPath checks whether the request URI path hierarchically matches the given path.
+// Examples:
+// / matches /
+// /foo matches /
+// /foo/ matches /
+// /foo/bla matches /
+// /foo/bla does not match /bla
+func matchesPath(requestURI string, path string) bool {
+	if path == "/" {
+		return true
+	}
+	if !strings.HasSuffix(requestURI, "/") {
+		requestURI += "/"
+	}
+	if !strings.HasSuffix(path, "/") {
+		path += "/"
+	}
+	return requestURI == path || strings.HasPrefix(requestURI, path)
+}
+
+func (h Engine) applyGlobalMiddleware(echoServer core.EchoRouter, serverConfig core.ServerConfig) {
+	// Use middleware to decode URL encoded path parameters like did%3Anuts%3A123 -> did:nuts:123
+	echoServer.Use(decodeURIPath)
+
+	skipper := func(c echo.Context) bool {
+		return matchesPath(c.Request().RequestURI, "/metrics") || matchesPath(c.Request().RequestURI, "/status")
+	}
+	echoServer.Use(loggerMiddleware(loggerConfig{Skipper: skipper, logger: log.Logger()}))
+
+	// Always enabled in strict mode
+	if serverConfig.Strictmode || serverConfig.InternalRateLimiter {
+		echoServer.Use(newInternalRateLimiter(map[string][]string{
+			http.MethodPost: {
+				"/internal/vcr/v2/issuer/vc",                   // issuing new VCs
+				"/internal/vdr/v1/did",                         // creating new DIDs
+				"/internal/vdr/v1/did/:did/verificationmethod", // add VM to DID
+				"/internal/didman/v1/did/:did/endpoint",        // add endpoint to DID
+				"/internal/didman/v1/did/:did/compoundservice", // add compound service to DID
+			},
+			http.MethodPut: {
+				"/internal/vdr/v1/did/:did",                // updating DIDs
+				"/internal/didman/v1/did/:did/contactinfo", // updating contactinfo in DID
+			}}, 24*time.Hour, 3000, 30),
+		)
+	}
+}
+
+func (h Engine) applyBindMiddleware(echoServer EchoServer, path string, excludePaths []string, serverConfig core.ServerConfig, cfg InterfaceConfig) error {
```

Comment:
```yaml
http:
  default:
    address: :1323
  alt:
    internal: # auth config is NOT accepted when address is not specified. /internal does not require a token with this config
      auth:
        type: token
    n2n: # does require token when address is specified
      address: :1323
      auth:
        type: token
```
The config above does not require a token when `http.alt.internal.auth` is set but `http.alt.internal.address` is not. If this is expected behavior it should be clearly documented.